File: Mathematics/Prefix Sum/atcoder/B06/B06.go
// ### 🔍 よくあるランタイムエラーの原因（Go）

// | 原因               | 内容                                                      |
// | ---------------- | ------------------------------------------------------- |
// | ❌ インデックスエラー      | `acc[L-1]` のように `L=1` のとき `acc[0]` にアクセスできない or スライス外参照 |
// | ❌ スキャン不足         | 入力のスキャン回数が `Q` に足りていない（クエリが途中で終わる）                      |
// | ❌ スペースのない入力の処理ミス | `bufio.Scanner` が 1行ずつ読み取る前提で、改行の無い入力に対応できていない          |
// | ❌ 空文字列の変換        | `strconv.Atoi("")` のように、空の文字列を数値に変換してエラー                |

// ### ✅ 修正版 Go コード（安全対策込み）

package main

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)

	// Read N
	line, _ := reader.ReadString('\n')
	N, _ := strconv.Atoi(strings.TrimSpace(line))

	// Read A
	line, _ = reader.ReadString('\n')
	Astr := strings.Fields(line)
	A := make([]int, N+1) // 1-indexed
	for i := 1; i <= N; i++ {
		A[i], _ = strconv.Atoi(Astr[i-1])
	}

	// Build prefix sum
	acc := make([]int, N+1)
	for i := 1; i <= N; i++ {
		acc[i] = acc[i-1] + A[i]
	}

	// Read Q
	line, _ = reader.ReadString('\n')
	Q, _ := strconv.Atoi(strings.TrimSpace(line))

	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	// Read and process queries
	for i := 0; i < Q; i++ {
		line, err := reader.ReadString('\n')
		if err != nil {
			break // 安全策
		}
		LR := strings.Fields(line)
		if len(LR) < 2 {
			continue // 不正入力回避
		}
		L, _ := strconv.Atoi(LR[0])
		R, _ := strconv.Atoi(LR[1])

		if L < 1 || R > N || L > R {
			writer.WriteString("draw\n") // 安全なフォールバック
			continue
		}

		ones := acc[R] - acc[L-1]
		total := R - L + 1
		zeros := total - ones

		if ones > zeros {
			writer.WriteString("win\n")
		} else if zeros > ones {
			writer.WriteString("lose\n")
		} else {
			writer.WriteString("draw\n")
		}
	}
}

// ## ✅ 修正ポイントまとめ

// | 修正箇所                        | 説明                      |   |                         |
// | --------------------------- | ----------------------- | - | ----------------------- |
// | ✅ `bufio.NewReader` 使用      | `bufio.Scanner` の行制限を回避 |   |                         |
// | ✅ `strings.Fields` 使用       | 空白区切り入力を安全に処理           |   |                         |
// | ✅ `acc` と `A` を 1-indexed に | `acc[L-1]` アクセスを安全にするため |   |                         |
// | ✅ クエリ境界チェック                 | \`L < 1                 |   | R > N\` などの入力ミスで落ちないように |

// ---

// ## ✅ 確認テスト（例）

// ### 入力

// ```
// 7
// 0 1 1 0 1 0 0
// 3
// 2 5
// 2 7
// 5 7
// ```

// ### 出力

// ```
// win
// draw
// lose
// ```

// ## ✅ 補足

// * 万が一、`LR` の長さが不足する場合（不完全な入力など）にも対応。
// * クエリ数が `Q` 未満でもクラッシュしないように `err != nil` チェック。

File: Mathematics/Prefix Sum/atcoder/B06/B06.php
<!-- 以下に、**型を明示した PHP 解法**を提示します。
この問題は高速な区間処理が必要なので、\*\*累積和（prefix sum）\*\*を用いて高速化します。

## ✅ PHP（型明示あり）完全解法

<?php

declare(strict_types=1);

/**
 * @param int $N
 * @param array<int> $A
 * @param array<array{int, int}> $queries
 * @return array<string>
 */
function solveLottery(int $N, array $A, array $queries): array {
    // 累積和 acc[i] = A[0] ~ A[i-1] の 1 の数（0-indexed）
    $acc = array_fill(0, $N + 1, 0); // int[]

    for ($i = 0; $i < $N; $i++) {
        $acc[$i + 1] = $acc[$i] + $A[$i];
    }

    $results = []; // array<string>

    foreach ($queries as [$L, $R]) {
        $ones = $acc[$R] - $acc[$L - 1];
        $total = $R - $L + 1;
        $zeros = $total - $ones;

        if ($ones > $zeros) {
            $results[] = "win";
        } elseif ($zeros > $ones) {
            $results[] = "lose";
        } else {
            $results[] = "draw";
        }
    }

    return $results;
}

// 標準入力処理
$lines = explode("\n", trim(file_get_contents("php://stdin")));

$N = (int)$lines[0];
$A = array_map('intval', explode(' ', $lines[1]));

$Q = (int)$lines[2];
$queries = []; // array<array{int, int}>
for ($i = 0; $i < $Q; $i++) {
    [$L, $R] = array_map('intval', explode(' ', $lines[3 + $i]));
    $queries[] = [$L, $R];
}

$results = solveLottery($N, $A, $queries);
echo implode("\n", $results) . "\n";

// ## ✅ 使用例（input）

// 7
// 0 1 1 0 1 0 0
// 3
// 2 5
// 2 7
// 5 7
// ```

// ## ✅ 実行

// ```bash
// php solution.php < input.txt
// ```

// ## ✅ 出力

// ```
// win
// draw
// lose
// ```

// ---

// ## ✅ 型注釈についての補足

// | 項目                  | 型                        | 意味                                     |
// | ------------------- | ------------------------ | -------------------------------------- |
// | `$N`                | `int`                    | くじの回数                                  |
// | `$A`                | `array<int>`             | くじ結果（0 または 1）                          |
// | `$queries`          | `array<array{int, int}>` | 区間クエリ `(L, R)` のリスト                    |
// | `solveLottery` の戻り値 | `array<string>`          | 各クエリに対する結果：`"win"`, `"lose"`, `"draw"` |

// ## ✅ 処理の流れまとめ

// 1. `$acc[i]` にアタリ（=1）の累積和を格納
// 2. 各クエリ `(L, R)` に対して `acc[R] - acc[L-1]` でアタリ数を高速取得
// 3. アタリ・ハズレを比較して出力

// ## ✅ 計算量

// * 累積和構築：`O(N)`
// * 各クエリ処理：`O(1)`
// * 合計：`O(N + Q)` → 最大 20万でも十分高速